python/lsst/daf/butler/datastores/fileLikeDatastore.py

```diff
@@ -847,7 +845,61 @@ def _read_artifact_into_memory(self, getInfo: DatastoreFileGetInformation,
         inMemoryDataset : `object`
             The artifact as a python object.
         """
-        raise NotImplementedError()
+        location = getInfo.location
+        uri = location.uri
+        log.debug("Accessing data from %s", uri)
+
+        # Cannot recalculate checksum but can compare size as a quick check
+        recorded_size = getInfo.info.file_size
+        resource_size = uri.size()
+        if resource_size != recorded_size:
+            raise RuntimeError("Integrity failure in Datastore. "
+                               f"Size of file {uri} ({resource_size}) "
+                               f"does not match size recorded in registry of {recorded_size}")
+
+        # For the general case we have choices for how to proceed.
+        # 1. Always use a local file (downloading the remote resource to a
+        #    temporary file if needed).
+        # 2. Use a threshold size and read into memory and use bytes.
+        # Use both for now with an arbitrary hand off size.
+        # This allows small datasets to be downloaded from remote object
+        # stores without requiring a temporary file.
+
+        formatter = getInfo.formatter
+        nbytes_max = 10_000_000  # Arbitrary number that we can tune
+        if resource_size <= nbytes_max and formatter.can_read_bytes():
+            serializedDataset = uri.read()
+            log.debug("Deserializing %s from %d bytes from location %s with formatter %s",
+                      f"component {getInfo.component}" if isComponent else "",
+                      len(serializedDataset), uri, formatter.name())
+            try:
+                result = formatter.fromBytes(serializedDataset,
+                                             component=getInfo.component if isComponent else None)
+            except Exception as e:
+                raise ValueError(f"Failure from formatter '{formatter.name()}' for dataset {ref.id}"
+                                 f" ({ref.datasetType.name} from {uri}): {e}") from e
+        else:
+            # Read from file
+            with uri.as_local() as local_uri:
+                # Have to update the Location associated with the formatter
+                # because formatter.read does not allow an override.
+                # This could be improved.
```

Comment:
FWIW, this is indeed something I'd love to see fixed (but not on this ticket).

Reply:
The FileDescriptor class was one of the first classes that Pim wrote and I'm not really convinced I need it. Again I can deal with that on the formatter cleanup.